kzg_prover/src/utils/amortized_kzg.rs

```diff
@@ -0,0 +1,147 @@
+use halo2_proofs::{
+    arithmetic::{best_fft, kate_division, Field},
+    halo2curves::{
+        bn256::{Bn256, Fr as Fp, G1Affine, G2Affine, Gt, G1},
+        group::{prime::PrimeCurveAffine, Curve, Group},
+        pairing::{Engine, PairingCurveAffine},
+    },
+    poly::{
+        commitment::{Blind, CommitmentScheme, Params, ParamsProver},
+        kzg::commitment::ParamsKZG,
+        Coeff, EvaluationDomain, Polynomial,
+    },
+};
+
+/// Commit to a polynomial using the KZG commitment scheme
+///
+/// # Arguments
+///
+/// * `params` - The KZG parameters
+/// * `poly` - The polynomial to commit to
+///
+/// # Returns
+///
+/// * The commitment to the polynomial as the G1 point
+pub fn commit_kzg(params: &ParamsKZG<Bn256>, poly: &Polynomial<Fp, Coeff>) -> G1 {
+    params.commit(poly, Blind::default())
+}
+
+/// Computes the polynomial h(X) for the amortized KZG algorithm as per FK23
+/// https://eprint.iacr.org/2023/033
+///
+/// # Arguments
+///
+/// * `params` - The KZG parameters
+/// * `f_poly` - The polynomial to compute the amortized opening proof for
+///
+/// # Returns
+///
+/// * The polynomial h(X) as a vector of G1 points
+pub fn compute_h(params: &ParamsKZG<Bn256>, f_poly: &Polynomial<Fp, Coeff>) -> Vec<G1> {
+    // Double the polynomial length, thus K + 1
+    let double_domain = EvaluationDomain::new(1, params.k() + 1);
+
+    let d = f_poly.len(); // Degree of the polynomial
+
+    // Extract s_commitments from ParamsKZG and extend with neutral elements
+    let mut s_commitments_reversed = params
+        .get_g()
+        .iter()
+        .map(PrimeCurveAffine::to_curve)
+        .collect::<Vec<_>>();
+    s_commitments_reversed.reverse();
+
+    let mut y: Vec<G1> = vec![G1::identity(); 2 * d];
+    y[..d].copy_from_slice(&s_commitments_reversed[..d]);
+
+    // Prepare coefficients vector and zero-pad at the beginning
+    let mut v = vec![Fp::zero(); 2 * d];
+    v[d..(2 * d)].copy_from_slice(f_poly);
+
+    let nu = double_domain.get_omega(); // 2d-th root of unity
+                                        // Perform FFT on s
+    best_fft(&mut y, nu, (2 * d).trailing_zeros());
+    // Perform FFT on c
```

Comment:
Does the "c" refer to "coefficients" of `v`?

Reply:
No, it's a naming mismatch between code and paper. The vector before FFT in the paper is named c and the result is named v, but our FFT function modifies the mutable vector in place. I'll unify the naming.

Reply:
I see, Thank you.